### ISSUE TITLE
Undo/Redo functionality

### ISSUE THREAD

**Issue by PlaylistsTrance** [Juna-Idler/LyricsTTE_Spectro#4]
Sometimes I accidentally overwrite a timestamp, because I forgot to go to the correct line. It would be amazing if I could use CTRL+Z to undo this change, and CTRL+Y to redo in case I just want to go back in history to check something I did, but want to keep my changes. The only actions that should be recorded are in TagStamp, TextEdit and TimeTune. Dragging the spectrogram should not be recorded for undoing, but changing timestamps in TagStamp, changing text in TextEdit and dragging timestamps in TimeTune should.

**Comment by Juna-Idler** [comment]
I think it would be convenient, but I think I don't use it much.

In Stamp mode, you only have to process it once and it will be completed, so there is not much need to overwrite it.
In Edit mode, it can be used with the basic specifications of the text box.
In Tune mode, you can just manually adjust it back.

The motivation is not high because the effort to implement Undo / Redo is quite large compared to the estimated usage frequency.


あれば便利だとは思うけれど、あまり使わないんじゃないかと。

Stampモードでは一度だけ処理すれば完了となるので、上書きする必要性があまりない。
Editモードではテキストボックスの基本仕様で使える。
Tuneモードではそのまま手動調整で状態を戻せばいい。

推定する利用頻度に対して、Undo/Redoの実装の手間が結構大きいのでモチベーションは高くないです。

**Comment by PlaylistsTrance** [comment]
You're right that you're only supposed to time once in stamp mode, but I don't really use tune mode. I prefer getting the right timestamp already in Stamp mode, by dragging the spectogram and pressing enter once the white line is at the point I want to time. But maybe I should just use your intended use case, which is to first let the song play once while making approximate stamps, and then correct them in Tune mode.

**Comment by Juna-Idler** [comment]
Speaking of which, there is a demo video of RythmiKaRuTTE, but I didn't make a demo video of LyricsTTE Web.
For the time being, a demo of the Windows version.
I haven't done it at all for a while, so the operation isn't sophisticated.
https://www.youtube.com/watch?v=--uiHplEEKU&t=499s
Originally, the image is auxiliary, and the main purpose of Tune mode (and this app) is to distinguish by ear.

そういえばRythmiKaRuTTEのデモ動画はあるけど、LyricsTTEWebのデモ動画一つも作ってなかった。
とりあえずWindows版のデモ。
しばらく全然やってなかったので、操作が洗練されてないけど。
https://www.youtube.com/watch?v=--uiHplEEKU&t=499s
もともと画像は補助で、耳で判別するのがTuneモードの（そしてこのアプリの）主目的。

**Comment by PlaylistsTrance** [comment]
This is my workflow if you're interested: https://youtu.be/cCjYkfG9fJY

**Comment by Juna-Idler** [comment]
If you do this, as I expected
Roughly align in Stamp mode and fine-tune in Tune mode.
Isn't it better?

このやり方をするなら、想定通りに、
Stampモードで大まかな位置を合わせて、Tuneモードで微調整。
の方がよくないですか？

**Comment by PlaylistsTrance** [comment]
To be honest, I haven't tried it yet because it wasn't necessary when doing it like this. I'll try doing it the intended way and see if I prefer it.
